rpcs3/Emu/Cell/lv2/sys_rsx.cpp

```diff
@@ -437,13 +468,33 @@ error_code sys_rsx_context_attribute(u32 context_id, u32 package_id, u64 a3, u64
 	{
 	case 0x001: // FIFO
 	{
-		rsx::eng_lock rlock(render);
 		const u64 get = static_cast<u32>(a3);
 		const u64 put = static_cast<u32>(a4);
-		vm::_ref<atomic_be_t<u64>>(render->dma_address + ::offset32(&RsxDmaControl::put)).release(put << 32 | get);
-		render->fifo_ctrl->set_get(static_cast<u32>(get));
-		render->last_known_code_start = get;
-		render->sync_point_request.release(true);
+		const u64 get_put = put << 32 | get;
+		{
+			rsx::eng_lock rlock(render);
+			std::lock_guard lock(render->sys_rsx_mtx);
+			render->new_get_put = get_put;
+		}
+
+		while (render->new_get_put == get_put)
+		{
+			if (Emu.IsStopped())
+			{
+				// Abort
+				if (render->new_get_put.compare_and_swap_test(get_put, u64{umax}))
```

Comment:
Why modify the value again? The code is simpler on both sides if you keep the producer/consumer boundary clean. Taking back the value here doesn't stop the FIFO from having read it already anyway.
Just set and wait here (with exit check) and simply read and modify on the FIFO side.

Reply:
Also, isn't this loop dangerous? You're not holding any lock during the check, another thread can deadlock this first one if the same value is written twice. Unrealistic of course, but it's easy to handle by simply not stopping to check. I still don't see why waiting is needed, it doesn't seem to achieve anything afaict.

Reply:
The waiting is for the GET/PUT values to be read as equal to the newest values written by the modifiing thread.
Also, abort is for complete state for savestates, either the operation is complete or need to be redone on savestate load.